core/src/main/java/com/scalar/db/common/error/CoreError.java

```diff
@@ -0,0 +1,215 @@
+      SubCategory.ILLEGAL_ARGUMENT,
+      "009",
+      "Cross-partition scan filtering is not enabled. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_PROJECTION(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "010",
+      "The specified projection is not found. Invalid projection: %s, Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_CLUSTERING_KEY_BOUNDARY(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "011",
+      "The clustering key boundary is not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_START_CLUSTERING_KEY(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "012",
+      "The start clustering key is not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_END_CLUSTERING_KEY(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "013",
+      "The end clustering key is not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_ORDERING1(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "014",
+      "Orderings are not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_ORDERING2(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "015",
+      "The specified ordering column is not found. Invalid ordering: %s, Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_CONDITION(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "016",
+      "The condition is not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_TABLE_NOT_FOUND(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "017",
+      "The specified table is not found: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_INVALID_COLUMN(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "018",
+      "The column value is not properly specified. Invalid column: %s, Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_MUTATIONS_EMPTY(
+      Category.USER_ERROR, SubCategory.ILLEGAL_ARGUMENT, "019", "The mutations are empty", "", ""),
+  OPERATION_CHECK_ERROR_MULTI_PARTITION_MUTATION(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "020",
+      "Mutations that span multi-partition are not supported. Mutations: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_PARTITION_KEY(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "021",
+      "The partition key is not properly specified. Operation: %s",
+      "",
+      ""),
+  OPERATION_CHECK_ERROR_CLUSTERING_KEY(
+      Category.USER_ERROR,
+      SubCategory.ILLEGAL_ARGUMENT,
+      "022",
+      "The clustering key is not properly specified. Operation: %s",
+      "",
+      ""),
+  ;
+
+  private static final String COMPONENT_NAME = "CORE";
+
+  private final Category category;
```

Comment:
`SubCategory` contains `Category`, so I think this class doesn't need to have `Category` so that we can simplified `ScalarDbError.validate()`. Maybe only holding `SubCategory` sounds a bit odd. If so, renaming `SubCategory` to `Type` or something might be an option?

Reply:
Yes, `SubCategory` includes `Category` as its `parent` to represent the parent-child relationship between `Category` and `SubCategory`. And, in the current implementation, the `parent` is used only for validation purposes.

I think it's natural that the error has a category as well as a sub-category, but the challenge is how to effectively model this relationship in Java.

> Maybe only holding SubCategory sounds a bit odd. If so, renaming SubCategory to Type or something might be an option?

I think only holding SubCategory sounds a bit odd, too. Instead of renaming `SubCategory` to `Type`, maybe we could perhaps a new enum holding both `Category` and `SubCategory` to represent the parent-child relationship. And in this case, we can remove `parent` from `SubCategory`. But I feel that it might be too much, so I'm not sure which is better the original one or this one.